src/sbd-cluster.c

```diff
@@ -67,21 +70,130 @@ sbd_plugin_membership_dispatch(cpg_handle_t handle,
+    }
+}
+
 void
 sbd_cpg_membership_dispatch(cpg_handle_t handle,
                     const struct cpg_name *groupName,
                     const struct cpg_address *member_list, size_t member_list_entries,
                     const struct cpg_address *left_list, size_t left_list_entries,
                     const struct cpg_address *joined_list, size_t joined_list_entries)
 {
-    if(member_list_entries > 0) {
-        set_servant_health(pcmk_health_online, LOG_INFO,
-                           "Connected to %s", name_for_cluster_type(get_cluster_type()));
+    cpg_membership_entries = member_list_entries;
+    sbd_cpg_membership_health_update();
+    notify_parent();
+}
+
+static void sbd_cmap_notify_fn(
+    cmap_handle_t cmap_handle,
+    cmap_track_handle_t cmap_track_handle,
+    int32_t event,
+    const char *key_name,
+    struct cmap_notify_value new_val,
+    struct cmap_notify_value old_val,
+    void *user_data)
+{
+    if ((event == CMAP_TRACK_MODIFY) &&
+       (new_val.type == CMAP_VALUETYPE_UINT8)) {
+        two_node = *((uint8_t *) new_val.data);
+        sbd_cpg_membership_health_update();
+        notify_parent();
+    }
+}
+
+static gboolean
+cmap_dispatch_callback (gpointer user_data)
+{
+    cmap_dispatch(cmap_handle, CS_DISPATCH_ALL);
+    return TRUE;
+}
+
+static gboolean
+sbd_get_two_node(void)
+{
+    uint8_t two_node_u8 = 0;
+    int cmap_fd;
+
+    if (!track_handle) {
+        if (cmap_initialize(&cmap_handle) != CS_OK) {
+            cl_log(LOG_WARNING, "Cannot initialize CMAP service\n");
+            goto out;
+        }
+
+        if (cmap_track_add(cmap_handle, "quorum.two_node", CMAP_TRACK_MODIFY,
+                            sbd_cmap_notify_fn, NULL, &track_handle) != CS_OK) {
+            cl_log(LOG_WARNING, "Failed adding CMAP tracker for 2Node-mode\n");
+            goto out;
+        }
+
+        /* add the tracker to mainloop */
+        cmap_fd_get(cmap_handle, &cmap_fd);
+        cmap_source = g_unix_fd_source_new (cmap_fd, G_IO_IN);
+        g_source_set_callback (cmap_source, cmap_dispatch_callback, NULL, NULL);
+        g_source_attach (cmap_source, NULL);
+    }
+
+    if (cmap_get_uint8(cmap_handle, "quorum.two_node", &two_node_u8) == CS_OK) {
+        cl_log(LOG_NOTICE, "Corosync is%s in 2Node-mode", two_node_u8?"":" not");
+        two_node = two_node_u8;
+
+        return TRUE;
     } else {
-        set_servant_health(pcmk_health_unclean, LOG_WARNING,
-                           "Empty %s membership", name_for_cluster_type(get_cluster_type()));
+        cl_log(LOG_WARNING, "Cannot get quorum.two_node from cmap-service\n");
     }
-    notify_parent();
+
+out:
+    if (cmap_source) {
+        g_source_destroy(cmap_source);
+    }
+    if (track_handle) {
+        cmap_track_delete(cmap_handle, track_handle);
+    }
+    if (cmap_handle) {
+        cmap_finalize(cmap_handle);
+    }
+    track_handle = 0;
+    cmap_handle = 0;
+    cmap_source = NULL;
```

Comment:
Depends on the taste, but would zero the variables along with respective destructors, for more clarity/less redundancy.